File: Algorithms/Math/0050_pow(x,n).py
#math
class Solution:
    def myPow(self, x: float, n: int) -> float:
        #x^n = x^n(1/2) * x^n(1/2) for Olog(n)
        #res = x ^ n(1/2)
        #if n is even, res = res * res, if n is odd, res = res * res * x
        #if n is < 1, return 1/ans
        #recursive loop 
        if n == 0: 
            return 1
        if x == 0: 
            return 0

        def calc(base, exp):
            if exp == 0:
                return 1
            res = calc(base, exp // 2)
            res = res * res 
            if exp % 2 == 1:
                res = res * base
            return res

        ans = calc(x,abs(n))
        if n < 0:
            return 1/ans 
        return ans

File: Algorithms/Math/0050_pow(x,n).c
// Practice for: 0050_pow(x,n).py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

double calc(double base, long long exp){
    if(exp == 0){
        return 1.0;
    }
    double res = calc(base, exp / 2);
    if (exp % 2 == 1){
        return res = res * res * base;
    } else {
        return res * res;
    } 
}

double myPow(double x, int n){
    long long N = n;
    long long ans = calc(x, abs(N));
    return n > 0 ? ans : 1/ans;
}
